src/common/hist_util.h

```diff
@@ -190,6 +190,101 @@ size_t DeviceSketch(int device,
                     DMatrix* dmat,
                     HistogramCuts* hmat);
 
+
+enum BinBounds {
+  UINT8_BINS_TYPE = 0,
+  UINT16_BINS_TYPE,
+  UINT32_BINS_TYPE,
+};
+
+struct Index {
+  Index(): binBound_(UINT8_BINS_TYPE), p_(1), disp_ptr_(nullptr) {
+    setBinBound(binBound_);
+  }
+  Index(const Index& i) = delete;
+  Index& operator=(Index i) = delete;
+  Index(Index&& i) = delete;
+  Index& operator=(Index&& i) = delete;
+  uint32_t operator[](size_t i) const {
+    if (disp_ptr_ != nullptr) {
+      return func_(data_ptr_, i) + disp_ptr_[i%p_];
+    } else {
+      return func_(data_ptr_, i);
+    }
+  }
+  void setBinBound(BinBounds binBound) {
+    binBound_ = binBound;
+    switch (binBound) {
+      case UINT8_BINS_TYPE:
+        func_ = &getValueFromUint8;
+        break;
+      case UINT16_BINS_TYPE:
+        func_ = &getValueFromUint16;
+        break;
+      case UINT32_BINS_TYPE:
+        func_ = &getValueFromUint32;
```

Comment:
This looks a lot like hand rolled polymorphism.

Reply:
It could be replaced on virtual function of course, but seems such approach is faster as there is no usage of VTable in each call of operator[] (Or just usage of switch operator would be applicable also).
This function is only needed for operator[]. And it's rarely used.